clients/go/admin/pkce/auth_flow_orchestrator.go

```diff
@@ -0,0 +1,166 @@
+package pkce
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"github.com/pkg/browser"
+	"golang.org/x/oauth2"
+)
+
+const (
+	Timeout     = 15 * time.Second
+	RefreshTime = 5 * time.Minute
+)
+
+// TokenOrchestrator implements the main logic to initiate Pkce flow to issue access token and refresh token as well as
+// refreshing the access token if a refresh token is present.
+type TokenOrchestrator struct {
+	clientConfig *oauth2.Config
+	tokenCache   TokenCache
+}
+
+// RefreshToken attempts to refresh the access token if a refresh token is provided.
+func (f TokenOrchestrator) RefreshToken(ctx context.Context, token *oauth2.Token) (*oauth2.Token, error) {
+	ts := f.clientConfig.TokenSource(ctx, token)
+	var refreshedToken *oauth2.Token
+	var err error
+	refreshedToken, err = ts.Token()
+	if err != nil {
+		logger.Warnf(ctx, "failed to refresh the token due to %v and will be doing re-auth", err)
+		return nil, err
+	}
+
+	if refreshedToken != nil {
+		logger.Debugf(ctx, "got a response from the refresh grant for old expiry %v with new expiry %v",
+			token.Expiry, refreshedToken.Expiry)
+		if refreshedToken.AccessToken != token.AccessToken {
+			if err = f.tokenCache.SaveToken(refreshedToken); err != nil {
+				logger.Errorf(ctx, "unable to save the new token due to %v", err)
+				return nil, err
+			}
+		}
+	}
+
+	return refreshedToken, nil
+}
+
+// FetchTokenFromCacheOrRefreshIt Fetch token from cache or refresh it. ignore the error while fetching from cache
+func (f TokenOrchestrator) FetchTokenFromCacheOrRefreshIt(ctx context.Context) (token *oauth2.Token, err error) {
+	token, err = f.tokenCache.GetToken()
+	if err != nil {
+		return nil, err
+	}
+
+	if !token.Valid() {
+		return nil, fmt.Errorf("token from cache is invalid")
+	}
+
+	// If token doesn't need to be refreshed, return it.
+	if token.Expiry.Add(RefreshTime).Before(time.Now()) {
+		return token, nil
+	}
+
+	token, err = f.RefreshToken(ctx, token)
+	if err != nil {
+		return nil, fmt.Errorf("failed to refresh token using cached token. Error: %w", err)
+	}
+
+	if !token.Valid() {
```

Comment:
what if token is nil?

Reply:
Valid() checks for that... nifty (or hacky?) go feature..

Reply:
oh! TIL